pkg/kubelet/kubelet_pods.go

```diff
@@ -1245,6 +1254,36 @@ func (kl *Kubelet) HandlePodCleanups(ctx context.Context) error {
 	return nil
 }
 
+// Filters terminal pods with deletion timestamp which are not runtime pods
+func (kl *Kubelet) filterTerminalPodsToDelete(allPods []*v1.Pod, runningRuntimePods []*kubecontainer.Pod, workingPods map[types.UID]PodWorkerSync) map[types.UID]*v1.Pod {
+	terminalPodsToDelete := make(map[types.UID]*v1.Pod)
+	for _, pod := range allPods {
+		if pod.DeletionTimestamp == nil {
+			// skip pods which don't have a deletion timestamp
+			continue
+		}
```

Comment:
Technically, we already have this check in canBeDeleted - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L897-L899.

However, if we remove this check here, we may trigger calling `TerminatePod` multiple times for pods that don't have a deletion timestamp set. And it looks like we pass `forceUpdate=true` here [1], so we will always construct a new status. So I suppose it does make sense to keep this check here unless we want refactor `TerminatePod` to short circuit if there is no update, but that is a more invasive change.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L489

Reply:
Yes, the idea is to avoid calling `TerminatePod` repeatedly without end (pod deletion). Still, it is possible that `TerminatePod` is called multiple times before the pod is deleted (if the deletion takes long, but this operation should generally be very fast).

Reply:
ack, that makes sense. Even if the termination takes longer, TerminatePod is re-entrant, so it should be fine.

Reply:
I don't think there is an issue from performance pov. First, the path is only invoked for terminal pods with the deletionTimestamp. The only scenario that produces such pods is described in the bug and is when the pod is non-admissible, and deleted during kubelet restart. So, it seems very unlikely to have many of such pods. Second, the TerminatePod is very fast as there is no killing of containers. I this the only reason it could take close to 2s is when the Delete request is throttled by the QPS or P&F at the API server side. However, since 1.27 the QPS for kubelet is 50, so we should be able to delete around 100 such pods in the 2s interval.

Reply:
My concern here is that `TerminatePod` does call `updateStatusInternal` with `forceUpdate=true` (https://github.com/kubernetes/kubernetes/blob/2802bbce14ca32cddbd0d0567cd33fb435f445e6/pkg/kubelet/status/status_manager.go).

Since `forceUpdate=true`, we will skip [this](https://github.com/kubernetes/kubernetes/blob/2802bbce14ca32cddbd0d0567cd33fb435f445e6/pkg/kubelet/status/status_manager.go#L657-L660), and always create a new `versionedPodStatus` and trigger `syncBatch`. In `syncBatch` we send request to API-server which can take some time (throttling, P&F, etc). And since this is part of the `HandlePodCleanups` call, the time spent in `syncBatch` will block the main sync loop.

Another alternative would be to refactor `TerminatePod` and have a way to mark the pod as terminated asynchronously so we don't trigger `forceUpdate=true` and have the pod termination be handled in the next `syncBatch`, that way it won't block the main `HandlePodCleanup` loop.

Reply:
Thanks for pointing this out. I will think about how to make the refactoring small enough for cherry-pick. 

Maybe this is a level of complication at which it is just simpler to add rejected pods to pod workers as proposed here: https://github.com/kubernetes/kubernetes/pull/118614/files. One downside (but seems really small) is that such pods are going `syncTerminatingPods` which might not be necessary. 

wdyt?

Reply:
Another alternative (suggested by @bobbypage ) is to call 
`kl.podWorkers.UpdatePod(UpdatePodOptions{ Pod: pod, UpdateType: kubetypes.SyncPodKill })` from within `HandlePodCleanups` instead of calling `TerminatePod`. It will allow us to skip the `UpdatePod` on the next `HandlePodCleanups` as it will be added to pod workers. I will give it a try.

Reply:
I've tested it and it works really well, updated the PR. It also eliminates the need for statusManager lookup. Moreover, we don't need to check the DeletionTimestamp, though it still might be desired.

Reply:
This is awesome, and I'm glad it worked out cleanly (also it makes handlepodcleanups even simpler and that's a huge win).

I think checking the deletion timestamp is still ok.

Reply:
+1, I think calling into the pod worker having the regular `SyncTerminatedPod` flow is much cleaner approach here, glad that is working out!